textdb/textdb-dataflow/src/main/java/com/google/re2j/PublicRegexp.java

```diff
@@ -0,0 +1,187 @@
+package com.google.re2j;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+/**
+ * Public Wrapper class for re2j.Regexp
+ * Regexp class represents the abstract syntax tree
+ * 
+ * @author Zuozhi Wang
+ *
+ */
+public class PublicRegexp extends Regexp {
+/*
+	// fields originally declared in Regexp:
+	// for detailed explanations please see getter methods
+	
+	// note that for variable subs, although original comments
+	// say it's never null, it could still be null
+  
+	Op op;                   // operator
+	int flags;               // bitmap of parse flags
+	Regexp[] subs;           // subexpressions, if any.  Never null.
+		                           // subs[0] is used as the freelist.
+	
+	int[] runes;             // matched runes, for LITERAL, CHAR_CLASS
+	int min, max;            // min, max for REPEAT
+	int cap;                 // capturing index, for CAPTURE
+	String name;             // capturing name, for CAPTURE
+	
+*/
+	
+	// publicSubs are an array of subexpressions with type PublicRegexp
+	private PublicRegexp[] publicSubs;
+
+	/**
+	 * private shallow copy constructor
+	 * shallow copy only copies reference to subexpression array
+	 */
+	private PublicRegexp(Regexp that) {
+		super(that);
+	}
+	  
+	/**
+	 * deep copy converts every Regexp Object in subexpressions to a PublicRegexp object
+	 * and put them in publicSubs array
+	 * @param Regexp re, A Regexp that needs to be converted to PublicRegexp
+	 * @return PublicRegex
+	 */
+	// 
+	public static PublicRegexp deepCopy(Regexp re) {
```

Comment:
I feel, it is better to move this method to some utils class (maybe PublicRegexpUtils or Re2jUtils), since PublicRegexp is a data object and it should represent only state instead of both behaviour and state